<?xml version="1.0" ?>
<launch>

  <node name="unspawn_arms"
        pkg="pr2_controller_manager" type="unspawner"
        args="l_arm_controller r_arm_controller" />


  <rosparam ns="r_cart">
    type: JTTeleopController
    root_name: torso_lift_link
    tip_name: r_gripper_tool_frame
    k_posture: 25.0
    jacobian_inverse_damping: 0.01
    pose_command_filter: 0.01
    cart_gains:
      trans:
        p: 1000.0
        d: 15.0
      rot:
        p: 90.0
        d: 1.2
    joint_feedforward:
      r_shoulder_pan_joint: 3.33
      r_shoulder_lift_joint: 1.16
      r_upper_arm_roll_joint: 0.1
      r_elbow_flex_joint: 0.25
      r_forearm_roll_joint: 0.133
      r_wrist_flex_joint: 0.0727
      r_wrist_roll_joint: 0.0727
    joint_max_effort:
      r_shoulder_pan_joint: 11.88
      r_shoulder_lift_joint: 11.64
      r_upper_arm_roll_joint: 6.143
      r_elbow_flex_joint: 6.804
      r_forearm_roll_joint: 8.376
      r_wrist_flex_joint: 5.568
      r_wrist_roll_joint: 5.568

    vel_saturation_trans: 1.0
    vel_saturation_rot: 2.0
  </rosparam>
  
  <rosparam ns="l_cart">
    type: JTTeleopController
    root_name: torso_lift_link
    tip_name: l_gripper_tool_frame
    k_posture: 25.0
    jacobian_inverse_damping: 0.01
    pose_command_filter: 0.01
    cart_gains:
      trans:
        p: 800.0
        d: 15.0
      rot:
        p: 80.0
        d: 1.2
    joint_feedforward:
      l_shoulder_pan_joint: 3.33
      l_shoulder_lift_joint: 1.16
      l_upper_arm_roll_joint: 0.1
      l_elbow_flex_joint: 0.25
      l_forearm_roll_joint: 0.133
      l_wrist_flex_joint: 0.0727
      l_wrist_roll_joint: 0.0727
    joint_max_effort:
      l_shoulder_pan_joint: 11.88
      l_shoulder_lift_joint: 11.64
      l_upper_arm_roll_joint: 6.143
      l_elbow_flex_joint: 6.804
      l_forearm_roll_joint: 8.376
      l_wrist_flex_joint: 5.568
      l_wrist_roll_joint: 5.568
    vel_saturation_trans: 1.0
    vel_saturation_rot: 2.0
  </rosparam>
  

  <node name="spawn_cart"
        pkg="pr2_controller_manager" type="spawner"
        args="l_cart r_cart" />

  <node name="r_posture"
        pkg="teleop_controllers" type="posture.py"
        args="r_cart elbowupr" />

  <node name="l_posture"
        pkg="teleop_controllers" type="posture.py"
        args="l_cart elbowupl" />


  <param name="success_threshold" value="0.02"/>
  <param name="max_lead" value="0.08"/>
  <param name="gripper_max_effort" value="18"/>
  <node name="grasp_server" pkg="recyclerbot" type="grasp_server" output="screen">
    <remap from="command_pose" to="/r_cart/command_pose"/>
    <remap from="current_pose" to="/r_cart/state/x"/>
  </node>
  
  <rosparam>
    type: PoseStamped
    object_pose:
      position:
        x: 0.84452944994
        y: 0.0548339411616
        z: 1.0111913681
      orientation:
        x: 0
        y: 0.3
        z: 0
        w: 0.953939201
    gripper_orientation:
      x: -0.00403457514937
      y: -0.780539478508
      z: 0.00205700227215
      w: -0.625090084256
    gripper_open_position: 0.08
    gripper_close_max_effort: 100.0
  </rosparam>
  
  <node name="object_detector" pkg="recyclerbot" type="object_detector_node" output="screen"/>
  
  <node name="recyclerbot" pkg="recyclerbot" type="recyclerbot_master" output="screen">
    <!--remap from="grasp_server" to="grasp_server"/-->
    <remap from="gripper_server" to="/r_gripper_controller/gripper_action"/>
  </node>
  
  
  <node name="r_gripper_controller"
  	pkg="pr2_controller_manager" type="pr2_controller_manager"
	args="start r_gripper_controller" />

  <node name="l_gripper_controller"
  	pkg="pr2_controller_manager" type="pr2_controller_manager"
	args="start l_gripper_controller" />
  
    
  <!--group ns="right_arm">
    <param name="success_threshold" value="0.02"/>
    <param name="max_lead" value="0.08"/>
    <node name="grasp_server" pkg="recyclerbot" type="grasp_server" output="screen">
      <remap from="command_pose" to="/r_cart/command_pose"/>
      <remap from="current_pose" to="/r_cart/state/x"/>
    </node>
  </group>

  <group ns="left_arm">
    <param name="success_threshold" value="0.02"/>
    <param name="max_lead" value="0.08"/>
    <node name="grasp_server" pkg="recyclerbot" type="grasp_server">
      <remap from="command_pose" to="/l_cart/command_pose"/>
      <remap from="current_pose" to="/l_cart/state/x"/>
    </node>
  </group>


  <group ns="right_arm">
    <group ns="chessboard_detector" clear_params="true">
      <param name="~corners_x" type="int" value="7"/>
      <param name="~corners_y" type="int" value="7"/>
      <param name="~publish_period" type="double" value="0.2"/>
      
      <node name="cb_detector" pkg="checkerboard_pose" type="cb_detector.py" respawn="false" output="screen">
	<remap from="camera_info" to="/narrow_stereo/left/camera_info"/>
	<remap from="image_stream" to="/narrow_stereo/left/image_rect"/>
	<remap from="~spacing_x" to="/right_chessbot/square_sidelength"/>
	<remap from="~spacing_y" to="/right_chessbot/square_sidelength"/>
      </node>
    </group>
    <node name="grasp_client" pkg="recyclerbot" type="grasp_client">
      <remap from="grasp_server" to="/right_arm/grasp_server"/>
      <remap from="gripper_server" to="/r_gripper_controller/gripper_action"/>
    </node>
    <rosparam>
      square_sidelength: 0.0339328125
      boardlevel_altitude: 0.07
      elevated_altitude: 0.12
      root_frame: torso_lift_link
      out_of_sight_pose:
        position:
          x: 0.842470855969
          y: -0.185501577783
          z: 0.315879531841
        orientation:
          x: -0.00403457514937
          y: -0.580539478508
          z: 0.00205700227215
          w: 0.814219506545
      gripper_orientation:
        x: -0.00403457514937
        y: -0.780539478508
        z: 0.00205700227215
        w: -0.625090084256
      gripper_open_position: 0.08
      gripper_close_max_effort: 100.0
    </rosparam>
  </group-->

  <!--group ns="left_chessbot">
    <group ns="chessboard_detector" clear_params="true">
      <param name="~corners_x" type="int" value="7"/>
      <param name="~corners_y" type="int" value="7"/>
      <param name="~publish_period" type="double" value="0.2"/>
      
      <node name="cb_detector" pkg="checkerboard_pose" type="cb_detector.py" respawn="false" output="screen">
	<remap from="camera_info" to="/narrow_stereo/left/camera_info"/>
	<remap from="image_stream" to="/narrow_stereo/left/image_rect"/>
	<remap from="~spacing_x" to="/left_chessbot/square_sidelength"/>
	<remap from="~spacing_y" to="/left_chessbot/square_sidelength"/>
      </node>
    </group>
    <node name="chessbot" pkg="chessbot" type="chessbot.py">
      <remap from="cartesian_arm_server" to="/left_arm/action_server"/>
      <remap from="gripper_server" to="/l_gripper_controller/gripper_action"/>
    </node>
    <rosparam>
      square_sidelength: 0.0339328125
      boardlevel_altitude: 0.07
      elevated_altitude: 0.12
      root_frame: torso_lift_link
      out_of_sight_pose:
        position:
          x: 0.828370973634
          y: 0.19033724064
          z: 0.310363133624
        orientation:
          x: -0.00403457514937
          y: -0.580539478508
          z: 0.00205700227215
          w: 0.814219506545
      gripper_orientation:
        x: -0.00403457514937
        y: -0.780539478508
        z: 0.00205700227215
        w: -0.625090084256
      gripper_open_position: 0.08
      gripper_close_max_effort: 100.0
    </rosparam>
  </group-->
</launch>
